src/main/scala/com/microsoft/hyperspace/index/execution/BucketUnionExec.scala

```diff
@@ -0,0 +1,112 @@
+ * retain outputPartitioning of result. So we define a new Union operation complying with
+ * the following conditions.
+ *   - input RDDs must have the same number of partitions.
+ *   - input RDDs must have the same partitioning keys.
+ *   - input RDDs must have the same column schema.
+ *
+ * Unfortunately, there is no explicit API for Partitioning keys in RDD, we have to
+ * assure that on the caller side. Therefore, [[BucketAwareUnionRDD]] is Hyperspace
+ * internal use only.
+ *
+ * You can find more detailed information about Bucketing optimization in
+ * [[https://databricks.com/session_na20/bucketing-2-0-improve-spark-sql-performance-by-removing-shuffle
+ Bucketing 2.0: Improve Spark SQL Performance by Removing Shuffle]]
+ */
+// scalastyle:on filelinelengthchecker
+private[hyperspace] class BucketAwareUnionRDD[T: ClassTag](
+    sc: SparkContext,
+    var rdds: Seq[RDD[T]],
+    bucketSpec: BucketSpec)
+    extends RDD[T](sc, rdds.map(x => new OneToOneDependency(x))) {
+  require(rdds.nonEmpty)
+  require(rdds.forall(_.getNumPartitions == bucketSpec.numBuckets))
+
+  // copy from org.apache.spark.rdd.PartitionerAwareUnionRDD
+  override def getPartitions: Array[Partition] = {
+    val numBuckets = bucketSpec.numBuckets
+    (0 until numBuckets).map { index =>
+      new BucketAwareUnionRDDPartition(rdds, index)
+    }.toArray
+  }
+
+  // copy from org.apache.spark.rdd.PartitionerAwareUnionRDD
+  override def compute(s: Partition, context: TaskContext): Iterator[T] = {
+    val parentPartitions = s.asInstanceOf[BucketAwareUnionRDDPartition].parents
+    rdds.zip(parentPartitions).iterator.flatMap {
+      case (rdd, p) => rdd.iterator(p, context)
+    }
+  }
+
+  // copy from org.apache.spark.rdd.PartitionerAwareUnionRDD
+  override def clearDependencies() {
+    super.clearDependencies()
+    rdds = null
+  }
+}
+
+/**
+ * [[BucketAwareUnionRDDPartition]] keeps partitions for each partition index.
+ * @param rdds  Input RDDs.
+ * @param index Partition index.
+ */
+private[hyperspace] class BucketAwareUnionRDDPartition(
+    @transient val rdds: Seq[RDD[_]],
+    override val index: Int)
+    extends Partition {
+  val parents: Array[Partition] = rdds.map(_.partitions(index)).toArray
+
+  override def hashCode(): Int = index
+  override def equals(other: Any): Boolean = super.equals(other)
+}
+
+/**
+ * [[BucketUnionExec]] is Spark Plan for [[BucketUnion]].
+ *
+ * @param children Child plans.
+ * @param bucketSpec Bucket specification.
+ */
+private[hyperspace] case class BucketUnionExec(children: Seq[SparkPlan], bucketSpec: BucketSpec)
+    extends SparkPlan {
+  override protected def doExecute(): RDD[InternalRow] = {
+    new BucketAwareUnionRDD[InternalRow](sparkContext, children.map(_.execute()), bucketSpec)
+  }
+
+  override def output: Seq[Attribute] = children.head.output
+  override def outputPartitioning: Partitioning = children.head.outputPartitioning
```

Comment:
Should we add a require for making sure all children have same output partitioning? If I understand correctly, this is the requirement check you mentioned above in the comments 
```
So we define a new Union operation complying with
 * the following conditions.
 *   - input RDDs must have the same number of partitions.
 *   - input RDDs must have the same partitioning keys.
 *   - input RDDs must have the same column schema.
 *
 * Unfortunately, there is no explicit API for Partitioning keys in RDD, we have to
 * assure that on the caller side
```
I could be wrong, please feel free to add comments and close this if it doesn't make sense.

Reply:
Suggestion was to add something like
`require(children.map(_.outputPartitioning).toSet.size == 1, "Children don't have same output partitioning")`

Reply:
@apoorvedave1 I tried to add this check but failed: 
https://github.com/microsoft/hyperspace/pull/123#discussion_r475366280

But I'll add some comment for this. Thanks!